strict mode?

Is there any feature I can turn on to require that an entire file is statically typed?

Or in general, what sort of tools would you use to enforce good mypy habits in a large project?

I would love to have this feature. Where would be a place to start? I'm completely unfamiliar with Mypy (but I am with Python).

The following is an example of what I would like to be warned on.

Here is correct code with type annotations:
```
def greeting() -> List[str]:
    return ["zapu"]

def main() -> None:
    print("Hello", ', '.join(greeting()))
```

Here is code that will fail type-check (and rightfully so):
```
def greeting() -> List[int]:
    return [1]

def main() -> None:
    print("Hello", ', '.join(greeting()))

output:

test.py: note: In function "main":
test.py:28: error: Argument 1 to "join" of "str" has incompatible type List[int]; expected Iterable[str]
```

Here is code that passes type-check but fails at runtime:
```
def greeting():
    return [1]

def main() -> None:
    print("Hello", ', '.join(greeting()))

output:

Traceback (most recent call last):
  File "test.py", line 32, in <module>
    main()
  File "test.py", line 28, in main
    print("Hello", ', '.join(greeting()))
TypeError: sequence item 0: expected str instance, int found
```

I am unsure what's the correct approach here, but consider that this could be avoided if type annotations were contagious. That is, every function that is type-annotated can only call type-annotated functions. So even if `greeting()` was in some other module, it would keep failing to type-check until I make a stub for it.

That sounds like the right approach to me.  I realized that that shouldn't be too hard to implement, so I took a stab at it above.  With #1230 (and `mypy --strict test.py`), here's what I get for your last example:

```python
def greeting():
    return [1]

def main() -> None:
    print("Hello", ', '.join(greeting()))
```
```
test.py: note: In function "main":
test.py:5: error: call to untyped function "greeting" in typed context
```

--- a/mypy/build.py
+++ b/mypy/build.py
@@ -51,7 +51,8 @@ TEST_BUILTINS = 'test-builtins'  # Use stub builtins to speed up tests
 DUMP_TYPE_STATS = 'dump-type-stats'
 DUMP_INFER_STATS = 'dump-infer-stats'
 SILENT_IMPORTS = 'silent-imports'  # Silence imports of .py files
-STRICT_MODE = 'strict-mode'      # Disallow calling untyped functions from typed ones
+# Disallow calling untyped functions from typed ones
+DISALLOW_UNTYPED_CALLS = 'disallow-untyped-calls'
 
 # State ids. These describe the states a source file / module can be in a
 # build.
@@ -365,7 +366,7 @@ class BuildManager:
         self.type_checker = TypeChecker(self.errors,
                                         modules,
                                         self.pyversion,
-                                        STRICT_MODE in self.flags)
+                                        DISALLOW_UNTYPED_CALLS in self.flags)
         self.states = []  # type: List[State]
         self.module_files = {}  # type: Dict[str, str]
         self.module_deps = {}  # type: Dict[Tuple[str, str], bool]

--- a/mypy/build.py
+++ b/mypy/build.py
@@ -51,6 +51,7 @@ TEST_BUILTINS = 'test-builtins'  # Use stub builtins to speed up tests
 DUMP_TYPE_STATS = 'dump-type-stats'
 DUMP_INFER_STATS = 'dump-infer-stats'
 SILENT_IMPORTS = 'silent-imports'  # Silence imports of .py files
+STRICT_MODE = 'strict-mode'      # Disallow calling untyped functions from typed ones
 
 # State ids. These describe the states a source file / module can be in a
 # build.
@@ -361,7 +362,10 @@ class BuildManager:
                                                   pyversion=pyversion)
         modules = self.semantic_analyzer.modules
         self.semantic_analyzer_pass3 = ThirdPass(modules, self.errors)
-        self.type_checker = TypeChecker(self.errors, modules, self.pyversion)
+        self.type_checker = TypeChecker(self.errors,
+                                        modules,
+                                        self.pyversion,
+                                        STRICT_MODE in self.flags)
         self.states = []  # type: List[State]
         self.module_files = {}  # type: Dict[str, str]
         self.module_deps = {}  # type: Dict[Tuple[str, str], bool]

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -364,12 +364,12 @@ class TypeChecker(NodeVisitor[Type]):
     # Have we deferred the current function? If yes, don't infer additional
     # types during this pass within the function.
     current_node_deferred = False
-    # Strict mode makes it an error to call an untyped function from a typed one
-    strict = False
+    # This makes it an error to call an untyped function from a typed one
+    disallow_untyped_calls = False
 
     def __init__(self, errors: Errors, modules: Dict[str, MypyFile],
                  pyversion: Tuple[int, int] = defaults.PYTHON3_VERSION,
-                 strict=False) -> None:
+                 disallow_untyped_calls=False) -> None:
         """Construct a type checker.
 
         Use errors to report type check errors. Assume symtable has been
@@ -393,7 +393,7 @@ class TypeChecker(NodeVisitor[Type]):
         self.deferred_nodes = []
         self.pass_num = 0
         self.current_node_deferred = False
-        self.strict = strict
+        self.disallow_untyped_calls = disallow_untyped_calls
 
     def visit_file(self, file_node: MypyFile, path: str) -> None:
         """Type check a mypy file with the given path."""

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -364,9 +364,12 @@ class TypeChecker(NodeVisitor[Type]):
     # Have we deferred the current function? If yes, don't infer additional
     # types during this pass within the function.
     current_node_deferred = False
+    # Strict mode makes it an error to call an untyped function from a typed one
+    strict = False
 
     def __init__(self, errors: Errors, modules: Dict[str, MypyFile],
-                 pyversion: Tuple[int, int] = defaults.PYTHON3_VERSION) -> None:
+                 pyversion: Tuple[int, int] = defaults.PYTHON3_VERSION,
+                 strict=False) -> None:
         """Construct a type checker.
 
         Use errors to report type check errors. Assume symtable has been
@@ -390,6 +393,7 @@ class TypeChecker(NodeVisitor[Type]):
         self.deferred_nodes = []
         self.pass_num = 0
         self.current_node_deferred = False
+        self.strict = strict
 
     def visit_file(self, file_node: MypyFile, path: str) -> None:
         """Type check a mypy file with the given path."""

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -139,7 +139,7 @@ class ExpressionChecker:
         # even if the type is known (in a dynamically typed function). This
         # way we get a more precise callee in dynamically typed functions.
         callee_type = self.chk.type_map[e.callee]
-        if (self.chk.strict and
+        if (self.chk.disallow_untyped_calls and
                 self.chk.typing_mode_full() and
                 isinstance(callee_type, CallableType)
                 and callee_type.implicit):

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -139,6 +139,11 @@ class ExpressionChecker:
         # even if the type is known (in a dynamically typed function). This
         # way we get a more precise callee in dynamically typed functions.
         callee_type = self.chk.type_map[e.callee]
+        if (self.chk.strict and
+                self.chk.typing_mode_full() and
+                isinstance(callee_type, CallableType)
+                and callee_type.implicit):
+            return self.msg.untyped_function_call(callee_type, e)
         return self.check_call_expr_with_callee_type(callee_type, e)
 
     # Types and methods that can be used to infer partial types.

--- a/mypy/main.py
+++ b/mypy/main.py
@@ -178,6 +178,9 @@ def process_options(args: List[str]) -> Tuple[List[BuildSource], Options]:
         elif args[0] == '--implicit-any':
             options.implicit_any = True
             args = args[1:]
+        elif args[0] == '--strict':
+            options.build_flags.append(build.STRICT_MODE)
+            args = args[1:]
         elif args[0] in ('--version', '-V'):
             ver = True
             args = args[1:]
@@ -307,6 +310,8 @@ Options:
   --py2              use Python 2 mode
   --python-version x.y  use Python x.y
   -s, --silent-imports  don't follow imports to .py files
+  --strict           disallow calling functions without type annotations from
+                     functions with type annotations
   --implicit-any     behave as though all functions were annotated with Any
   -f, --dirty-stubs  don't warn if typeshed is out of sync
   --pdb              invoke pdb on fatal error

--- a/mypy/main.py
+++ b/mypy/main.py
@@ -178,8 +178,8 @@ def process_options(args: List[str]) -> Tuple[List[BuildSource], Options]:
         elif args[0] == '--implicit-any':
             options.implicit_any = True
             args = args[1:]
-        elif args[0] == '--strict':
-            options.build_flags.append(build.STRICT_MODE)
+        elif args[0] == '--disallow-untyped-calls':
+            options.build_flags.append(build.DISALLOW_UNTYPED_CALLS)
             args = args[1:]
         elif args[0] in ('--version', '-V'):
             ver = True
@@ -310,8 +310,8 @@ Options:
   --py2              use Python 2 mode
   --python-version x.y  use Python x.y
   -s, --silent-imports  don't follow imports to .py files
-  --strict           disallow calling functions without type annotations from
-                     functions with type annotations
+  --disallow-untyped-calls  disallow calling functions without type annotations
+                            from functions with type annotations
   --implicit-any     behave as though all functions were annotated with Any
   -f, --dirty-stubs  don't warn if typeshed is out of sync
   --pdb              invoke pdb on fatal error

--- a/mypy/messages.py
+++ b/mypy/messages.py
@@ -391,6 +391,11 @@ class MessageBuilder:
         self.fail('{} not callable'.format(self.format(typ)), context)
         return AnyType()
 
+    def untyped_function_call(self, callee: CallableType, context: Context) -> Type:
+        name = callee.name if callee.name is not None else '(unknown)'
+        self.fail('call to untyped function {} in typed context'.format(name), context)
+        return AnyType()
+
     def incompatible_argument(self, n: int, m: int, callee: CallableType, arg_type: Type,
                               context: Context) -> None:
         """Report an error about an incompatible argument type.

--- a/mypy/nodes.py
+++ b/mypy/nodes.py
@@ -1761,6 +1761,7 @@ def function_type(func: FuncBase, fallback: 'mypy.types.Instance') -> 'mypy.type
             mypy.types.AnyType(),
             fallback,
             name,
+            implicit=True,
         )
 
 

--- a/mypy/types.py
+++ b/mypy/types.py
@@ -274,6 +274,8 @@ class CallableType(FunctionLike):
 
     # Is this Callable[..., t] (with literal '...')?
     is_ellipsis_args = False
+    # Was this type implicitly generated instead of explicitly specified by the user?
+    implicit = False
 
     def __init__(self, arg_types: List[Type],
                  arg_kinds: List[int],
@@ -285,7 +287,8 @@ class CallableType(FunctionLike):
                  variables: List[TypeVarDef] = None,
                  bound_vars: List[Tuple[int, Type]] = None,
                  line: int = -1,
-                 is_ellipsis_args: bool = False) -> None:
+                 is_ellipsis_args: bool = False,
+                 implicit=False) -> None:
         if variables is None:
             variables = []
         if not bound_vars:
@@ -303,6 +306,7 @@ class CallableType(FunctionLike):
         self.variables = variables
         self.bound_vars = bound_vars
         self.is_ellipsis_args = is_ellipsis_args
+        self.implicit = implicit
         super().__init__(line)
 
     def copy_modified(self,

